### PR TITLE
Issue106 reject on low memory

### PR DESCRIPTION
**Pull request by mcweba**

### CODE REVIEW

File: src/main/java/org/swisspush/redisques/util/DefaultMemoryUsageProvider.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+                    .setScale(2, RoundingMode.HALF_UP)
+                    .floatValue();
+
+            log.info("Current memory usage is {}%", currentMemoryUsagePercentage);
+            currentMemoryUsageOptional = Optional.of(currentMemoryUsagePercentage);
+        });
+    }
+
+
+    private Long totalSystemMemory(String memoryInfo) {
```
[the remark this answers is not in the document]
Comment (mcweba) [reply src/main/java/org/swisspush/redisques/util/DefaultMemoryUsageProvider.java]: Changed implementation to use Optional<Long>

File: src/main/java/org/swisspush/redisques/util/RedisquesConfiguration.java [hunk 2]
```diff
@@ -143,6 +143,13 @@ public RedisquesConfiguration(String address, String configurationUpdatedAddress
         this.queueConfigurations = queueConfigurations;
         this.enableQueueNameDecoding = enableQueueNameDecoding;
         this.queueSpeedIntervalSec = queueSpeedIntervalSec;
+
+        if(0 <= memoryUsageLimitPct && memoryUsageLimitPct <= 100){
```
Comment (dominik-cnx) [inline src/main/java/org/swisspush/redisques/util/RedisquesConfiguration.java:157]: code format

**Review by dominik-cnx — APPROVED** [review]

File: src/test/java/org/swisspush/redisques/handler/RedisquesHttpRequestHandlerTest.java [hunk 3]
```diff
@@ -240,14 +241,22 @@ public void getConfiguration(TestContext context) {
...
                 .body("any { it.key == 'checkInterval' }", is(true))
                 .body("any { it.key == 'processorTimeout' }", is(true))
                 .body("any { it.key == 'processorDelayMax' }", is(true))
+                .body("any { it.key == 'queueConfigurations' }", is(true))
+                .body("any { it.key == 'enableQueueNameDecoding' }", is(true))
+                .body("any { it.key == 'maxPoolSize' }", is(true))
+                .body("any { it.key == 'maxPoolWaitingSize' }", is(true))
+                .body("any { it.key == 'maxPipelineWaitingSize' }", is(true))
+                .body("any { it.key == 'queueSpeedIntervalSec' }", is(true))
+                .body("any { it.key == 'memoryUsageLimitPct' }", is(true))
```
[the remark this answers is not in the document]
Comment (mcweba) [reply src/test/java/org/swisspush/redisques/handler/RedisquesHttpRequestHandlerTest.java:258]: will apply this

File: README.md [hunk 4]
```diff
@@ -44,22 +44,30 @@ If an instance dies, its queues will be assigned to other instances.
...
+| maxPipelineWaitingSize        | 2048                            | The maximum allowed queued waiting handlers                                                                                                                                                     |
+| checkInterval                 | 60                              | The interval [s] to check timestamps of not-active / empty queues by executing **check** queue operation. _checkInterval_ value must be greater 0, otherwise the default is used.               |
+| queueSpeedIntervalSec         | 60                              | The interval [s] to check queue speed                                                                                                                                                           |
+| memoryUsageLimitPct           | 100                             | Percentage of the available system memory to be used by vertx-redisques. Only values between 0 and 100 are allowed. When the used memory ratio is higher than this limit, enqueues are rejected |
```
[the remark this answers is not in the document]
Comment (mcweba) [reply README.md:63]: will apply this

File: src/main/java/org/swisspush/redisques/util/DefaultMemoryUsageProvider.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+            return Optional.of(Long.parseLong(usedMemoryOpt.get().split(":")[1]));
+        } catch (NumberFormatException ex) {
+            logPropertyWarning("used_memory", ex);
+            return Optional.empty();
+        }
+    }
+
+    private void logPropertyWarning(String property, Exception ex) {
+        log.warn("No or invalid '{}' value received from redis. Unable to calculate the current memory usage. " +
+                "Exception: {}", property, ex.toString());
```
Comment (hiddenalpha) [inline src/main/java/org/swisspush/redisques/util/DefaultMemoryUsageProvider.java:111]:
```java
    ex.toString()
```
Why do we pass exception by calling "*.toString()*" and pass it via placeholder?

[Logging Exceptions Using SLF4J](https://www.baeldung.com/slf4j-log-exceptions)

**Pull request merged**